### PR TITLE
Chebyshev polynomial of the first kind

### PR DESCRIPTION
**Pull request by 0x00b1**
Adds:

```Python
chebyshev_polynomial_t(input, n, *, out=None) -> Tensor
```

Chebyshev polynomial of the first kind $T_{n}(\text{input})$.

If $n = 0$, $1$ is returned. If $n = 1$, $\text{input}$ is returned. If $n < 6$ or $|\text{input}| > 1$ the recursion:

$$T_{n + 1}(\text{input}) = 2 \times \text{input} \times T_{n}(\text{input}) - T_{n - 1}(\text{input})$$

is evaluated. Otherwise, the explicit trigonometric formula:

$$T_{n}(\text{input}) = \text{cos}(n \times \text{arccos}(x))$$

is evaluated.

## Derivatives

Recommended $k$-derivative formula with respect to $\text{input}$: 

$$2^{-1 + k} \times n \times \Gamma(k) \times C_{-k + n}^{k}(\text{input})$$ 

where $C$ is the Gegenbauer polynomial.

Recommended $k$-derivative formula with respect to $\text{n}$: 

$$\text{arccos}(\text{input})^{k} \times \text{cos}(\frac{k \times \pi}{2} + n \times \text{arccos}(\text{input})).$$

## Example

```Python
x = torch.linspace(-1, 1, 256)

matplotlib.pyplot.plot(x, torch.special.chebyshev_polynomial_t(x, 10))
```

image

### CODE REVIEW

File: aten/src/ATen/native/Math.h [hunk 1]
```diff
@@ -2177,4 +2177,48 @@ static inline C10_HOST_DEVICE T calc_log_ndtr(T x) {
...
+
+    if (n == 1) {
+        return x;
+    }
+
+    T p = T(1.0);
+    T q = x;
+    T r;
+
+    for (int k = 2; k <= n; k++) {
```
Comment (mruberry) [inline aten/src/ATen/native/Math.h:2210]: type k like n

File: aten/src/ATen/native/cuda/Math.cuh [hunk 2]
```diff
@@ -1265,6 +1265,52 @@ const auto erfcx_string = jiterator_stringify(
   }
 ); // erfcx_string
 
+const auto chebyshev_polynomial_t_string = jiterator_stringify(
+    template<typename T>
+    T chebyshev_polynomial_t_forward(T x, int n) {
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Math.cuh:1270]: explicit type for n

File: torch/special/__init__.py [hunk 3]
```diff
@@ -820,3 +851,30 @@
...
+chebyshev_polynomial_t = _add_docstr(_special.special_chebyshev_polynomial_t,
+                                     r"""
+chebyshev_polynomial_t(input, n, *, out=None) -> Tensor
+
+Chebyshev polynomial of the first kind :math:`T_{n}(\text{input})`.
+
+If :math:`n = 0`, :math:`1` is returned. If :math:`n = 1`, :math:`\text{input}` 
+is returned. If :math:`n < 6` or :math:`|\text{input}| > 1` the recursion:
+
+.. math::
```
Comment (mruberry) [inline torch/special/__init__.py:864]: Nice explanation -- be sure to check the formatting using the formatted doc link

File: torch/special/__init__.py [hunk 4]
```diff
@@ -820,3 +851,30 @@
...
+is evaluated. Otherwise, the explicit trigonometric formula:
+
+.. math::
+    T_{n}(\text{input}) = \text{cos}(n \times \text{arccos}(x))
+
+is evaluated.
+
+""" + r"""
+Args:
+    {input}
```
Comment (mruberry) [inline torch/special/__init__.py:876]: n is also an arg

**Review by mruberry — APPROVED** [review]
See a few small notes; otherwise once the tests pass this looks great!

**Comment by 0x00b1** [comment]
@jeffdaily Do you understand this `rocm` failure?

> error: reference to __host__ function 'chebyshev_polynomial_t_forward<double>' in __host__ __device__ function
    return chebyshev_polynomial_t_forward(x, static_cast<std::int64_t>(n));

File: aten/src/ATen/native/BinaryOps.cpp [hunk 5]
```diff
@@ -382,6 +391,22 @@ Tensor& special_zeta_out(const Tensor& self, const Scalar& other, Tensor& result
   return at::special_zeta_out(result, self, wrapped_scalar_tensor(other));
 }
 
+Tensor special_chebyshev_polynomial_t(const Scalar& x, const Tensor& n) {
```
Comment (ngimel) [inline aten/src/ATen/native/BinaryOps.cpp:394]: do you really want `n` argument to be a tensor of potentially same size as `x` and different `n` values, or is this just for convenience?

File: aten/src/ATen/native/Math.h [hunk 6]
```diff
@@ -2177,4 +2177,48 @@ static inline C10_HOST_DEVICE T calc_log_ndtr(T x) {
   }
 }
 
+template<typename T>
+static inline T chebyshev_polynomial_t_forward(T x, std::int64_t n) {
```
Comment (ngimel) [inline aten/src/ATen/native/Math.h:2181]: ROCM error is likely because this function is not marked C10_HOST_DEVICE

**Comment by 0x00b1** [comment]
@suo Should I be concerned about these failures? They don't appear to be relevant to this pull request.

File: aten/src/ATen/native/Math.h (hunk 6 above)
Comment (0x00b1) [reply aten/src/ATen/native/Math.h]: Thank you @ngimel! That resolved the issue. :)

File: aten/src/ATen/native/BinaryOps.cpp (hunk 5 above)
Comment (0x00b1) [reply aten/src/ATen/native/BinaryOps.cpp:394]: You might! I also wanted to match the API of the equivalent SciPy function (scipy.special.eval_chebyshevt).

Comment (0x00b1) [reply aten/src/ATen/native/BinaryOps.cpp:394]: The unusual case, in my opinion, is actually the `(const Tensor& x, const Tensor& n)` overload. I provided that for convenience.

**Comment by 0x00b1** [comment]
Awesome! Thanks @mruberry!

**Pull request closed without merging**